dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala

```diff
@@ -0,0 +1,176 @@
+import org.apache.spark.sql.types.StructType
+
+import org.apache.kyuubi.sql.KyuubiSQLConf
+
+/**
+ * Add maxPartitions Strategy to avoid scan excessive partitions on partitioned table
+ * 1 Check if scan exceed maxPartition
+ * 2 Check if Using partitionFilter on partitioned table
+ * This Strategy Add Planner Strategy after LogicalOptimizer
+ */
+case class MaxPartitionStrategy(session: SparkSession)
+  extends Strategy
+  with SQLConfHelper
+  with PruneFileSourcePartitionHelper {
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = {
+    conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS) match {
+      case Some(maxScanPartitions) => plan match {
+          case ScanOperation(_, _, relation: HiveTableRelation) if relation.isPartitioned =>
+            relation.prunedPartitions match {
+              case Some(prunedPartitions) =>
+                if (prunedPartitions.size > maxScanPartitions) {
+                  throw new MaxPartitionExceedException(
+                    s"""
+                       |SQL job scan hive partition: ${prunedPartitions.size}
+                       |exceed restrict of hive scan maxPartition $maxScanPartitions
+                       |You should optimize your SQL logical according partition structure
+                       |or shorten query scope such as p_date, detail as below:
+                       |Table: ${relation.tableMeta.qualifiedName}
+                       |Owner: ${relation.tableMeta.owner}
+                       |Partition Structure: ${relation.partitionCols.map(_.name).mkString(" -> ")}
+                       |""".stripMargin)
+                } else {
+                  Nil
+                }
+              case _ =>
+                val totalPartitions = session
+                  .sessionState.catalog.externalCatalog.listPartitionNames(
+                    relation.tableMeta.database,
+                    relation.tableMeta.identifier.table)
+                if (totalPartitions.size > maxScanPartitions) {
+                  throw new MaxPartitionExceedException(
+                    s"""
+                       |Your SQL job scan a whole huge table without any partition filter,
+                       |You should optimize your SQL logical according partition structure
+                       |or shorten query scope such as p_date, detail as below:
+                       |Table: ${relation.tableMeta.qualifiedName}
+                       |Owner: ${relation.tableMeta.owner}
+                       |Partition Structure: ${relation.partitionCols.map(_.name).mkString(" -> ")}
+                       |""".stripMargin)
+                } else {
+                  Nil
+                }
+            }
+          case ScanOperation(
+                _,
+                filters,
+                relation @ LogicalRelation(
+                  fsRelation @ HadoopFsRelation(
+                    fileIndex: InMemoryFileIndex,
```

Comment:
does this `InMemoryFileIndex` only comes from `PruneFileSourcePartitions` who's relation has `partitionKeyFilters` ?

Reply:
we need both to check a ScanOperation w/ or w/o partition key filters

- a ScanOperation with partition key filters

```sql
SELECT * FROM test where p = 1
```

partitionKeyFilters = ExpressionSet(isnotnull(p#45), (p#45 = 1))

only 1 partition will be scan

- a ScanOperation without partition key filters

```sql
SELECT * FROM test where
```
partitionKeyFilters = ExpressionSet()

as `partitionKeyFilters` is empty, `fileIndex.listFiles(partitionKeyFilters.toSeq, dataFilter)` will return all of partitions



So this strategy will be applied to all of the file-based partitioned tables.